src/group_theory/free_abelian_group_finsupp.lean

```diff
@@ -0,0 +1,146 @@
+/-
+Copyright (c) 2021 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+
+import group_theory.free_abelian_group
+import data.finsupp.basic
+
+/-!
+# Isomorphism between `free_abelian_group X` and `X →₀ ℤ`
+
+In this file we construct the canonical isomorphism between `free_abelian_group X` and `X →₀ ℤ`.
+We use this to transport the notion of `support` from `finsupp` to `free_abelian_group`.
+
+## Main declarations
+
+- `free_abelian_group.equiv_finsupp`: group isomorphism between `free_abelian_group X` and `X →₀ ℤ`
+- `free_abelian_group.coeff`: the multiplicity of `x : X` in `a : free_abelian_group X`
+- `free_abelian_group.support`: the finset of `x : X` that occur in `a : free_abelian_group X`
+
+-/
+
+noncomputable theory
+
+open_locale big_operators
+
+variables (X : Type*)
+
+/-- The group homomorphism `free_abelian_group X →+ (X →₀ ℤ)`. -/
+def free_abelian_group.to_finsupp : free_abelian_group X →+ (X →₀ ℤ) :=
+free_abelian_group.lift $ λ x, finsupp.single x (1 : ℤ)
+
+/-- The group homomorphism `(X →₀ ℤ) →+ free_abelian_group X`. -/
+def finsupp.to_free_abelian_group : (X →₀ ℤ) →+ free_abelian_group X :=
+finsupp.lift_add_hom $ λ x, (smul_add_hom ℤ (free_abelian_group X)).flip (free_abelian_group.of x)
+
+open finsupp free_abelian_group
+
+@[simp] lemma finsupp.to_free_abelian_group_comp_single_add_hom (x : X) :
+  (finsupp.to_free_abelian_group X).comp (finsupp.single_add_hom x) =
+    (smul_add_hom ℤ (free_abelian_group X)).flip (of x) :=
+begin
+  ext,
+  simp only [add_monoid_hom.coe_comp, finsupp.single_add_hom_apply, function.comp_app,
+    one_smul, to_free_abelian_group, finsupp.lift_add_hom_apply_single]
+end
+
+@[simp] lemma free_abelian_group.to_finsupp_comp_to_free_abelian_group :
+  (to_finsupp X).comp (to_free_abelian_group X) = add_monoid_hom.id _ :=
+begin
+  ext x y, simp only [add_monoid_hom.id_comp],
+  rw [add_monoid_hom.comp_assoc, finsupp.to_free_abelian_group_comp_single_add_hom],
+  simp only [to_finsupp, add_monoid_hom.coe_comp, finsupp.single_add_hom_apply,
+    function.comp_app, one_smul, lift.of, add_monoid_hom.flip_apply,
+    smul_add_hom_one, add_monoid_hom.id_apply],
+end
+
+@[simp] lemma finsupp.to_free_abelian_group_comp_to_finsupp :
+  (to_free_abelian_group X).comp (to_finsupp X) = add_monoid_hom.id _ :=
+begin
+  ext,
+  simp only [to_free_abelian_group, to_finsupp, finsupp.lift_add_hom_apply_single,
+    add_monoid_hom.coe_comp, function.comp_app, one_smul, add_monoid_hom.id_apply, lift.of,
+    add_monoid_hom.flip_apply, smul_add_hom_one],
+end
+
+@[simp] lemma finsupp.to_free_abelian_group_to_finsupp {X} (x : free_abelian_group X) :
+  (to_free_abelian_group X) (to_finsupp X x) = x :=
```

Comment:
```suggestion
  x.to_finsupp.to_free_abelian_group = x :=
```

Reply:
Wait? dot-notation doesn't work with bundled homs, does it?

Reply:
It does as long as they have no non-bundled arguments